Api/API Standardization.md

```diff
@@ -0,0 +1,305 @@
+    - **204 No Content** - if an existing resource was replaced
+    - **400 Bad Request** - if the action could not be completed due to an invalid request made by the client. Must contain the problem details in the response body.
+    - **401 Unauthorized**
+    - **403 Forbidden**
+
+## PATCH
+
+- By default, we **do not support** the PATCH method. This is because the behavior of this method is not strictly defined in the RFC standard, nor is there a commonly used standard for it.
+- We have previously used this method in the projects that used JSON:API standard, which explicitly does not support PUT, only PATCH.
+- If the project requires usage of a PATCH method, we recommend using the JSON PATCH standard which [.NET supports](https://learn.microsoft.com/en-us/aspnet/core/web-api/jsonpatch?view=aspnetcore-9.0). However, there are security implications with this standard the developers must address when using it.
+
+## DELETE
+
+```jsx
+Request:
+PUT www.example.com/users/4
+
+Response:
+204 No Content
+```
+
+- **Purpose:** To remove a specific resource.
+- **Safe**: No.
+- **Idempotent**: Yes - the response might not always be the same (subsequent requests might get a `404 Not Found` response), but the outcome on the server will be the same.
+- **Request body**: Should not have a request body.
+- **Possible responses:**
+    - **200 OK**
+    - **204 No Content**
+    - **400 Bad Request**
+    - **401 Unauthorized**
+    - **403 Forbidden**
+    - **404 Not Found**
+
+## Other common verbs
+
+### `HEAD`
+
+- **Purpose**: Identical to `GET`, but the server **must not** send a response body. It's used to check a resource's metadata (like `Content-Type` or `Last-Modified` headers) without downloading the entire resource.
+- **Safe**: **Yes**.
+- **Idempotent**: **Yes**.
+
+### `OPTIONS`
+
+- **Purpose**: To describe the communication options for the target resource. It's often used by browsers for CORS (Cross-Origin Resource Sharing) preflight requests to see which methods and headers are allowed by the server.
+- **Safe**: **Yes**.
+- **Idempotent**: **Yes**.
+
+## Returning the target resource in the response body
+
+The RFC standard does not specify if a created, updated, or deleted resource needs to be included in the response body. This decision is left to the developers to decide, depending on the project needs. Note that whatever decision is made on the project-level, **it must be consistent for all endpoints** - if the decision is made to return the resource, that must be done throughout the API, not just specific endpoints.
+
+We recommend that, by default, the API **does not** return the content it created or updated, although we do recommend returning the ID of the new resource if it was created as a result of a request. That being said, there are a lot of cases where the clients depend on the data returned to update their state, so be mindful of the API consumers when making that decision.
```

Comment:
I would default that we always return the  content. Since in 99% cases we already need to return the ID of newly created resource, I see no reason why we wouldn't return the whole DTO. Even in projects where returned data is not being used after POST (for example), it is likely that it will be used for PUT. That way we end up with mix of standards, and its sometimes hard to go back and update all to follow the same logic. For those reasons I'd advise returning the newly created/updated object always, and use 204 for edge cases where that is not possible.